File: Stack/min-stack/Python3/min-stack.py
'''
Description |   MinStack
            |   Time complexity for min, push pop are O(1)
------------+----------------------------------
Details     |   This python version is different with C++ version.  
            |   This version doesn't require additional space.
            |   It requires you to calculate the different when append element to the stack
------------+----------------------------------
Language    |   Python3
------------+----------------------------------
Leetcode    |   https://leetcode-cn.com/problems/bao-han-minhan-shu-de-zhan-lcof/
------------+----------------------------------
Compile     |   g++ -o min-stack min-stack.cpp
            |   ./min-stack
'''


class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minVal = -1


    def push(self, x: int) -> None:
        if not self.stack:
            self.stack.append(0) #first element is stored as 0
            self.minVal = x
        else:
            diff = x - self.minVal
            self.stack.append(diff)
            self.minVal = self.minVal if diff > 0 else x

    def pop(self) -> None:
        diff = self.stack.pop()
        if diff > 0:
            return diff+self.minVal
        else:
            self.minVal = self.minVal - diff
            return diff + self.minVal
        
    def top(self) -> int:
        diff = self.stack[-1]
        if diff < 0:
            return self.minVal
        return diff + self.minVal

    def min(self) -> int:
        return self.minVal


def main():
    stack = MinStack()
    stack.push(-2)
    stack.push(0)
    stack.push(-3)

    assert(stack.min() == -3)
    stack.pop()
    assert(stack.min() == -2)
    assert(stack.top() == 0)
    stack.pop()
    assert(stack.top() == -2)   
    print("Test Completed!\n");
    return 

if __name__ == "__main__":
    main()

File: Stack/min-stack/C++/min-stack.cpp
/**************************************************************************************************
Description |   MinStack
            |   Time complexity for min, push pop are O(1)
------------+----------------------------------
Details     |   Maintain 2 stacks 
            |   Sorted stack and Data Stack
------------+----------------------------------
Language    |   C++
------------+----------------------------------
Leetcode    |   https://leetcode-cn.com/problems/bao-han-minhan-shu-de-zhan-lcof/
------------+----------------------------------
Compile     |   g++ -o min-stack min-stack.cpp
            |   ./min-stack
**************************************************************************************************/
#include <cassert>
#include <cstdio>
#include <stack>  

using namespace std;
class MinStack {
public:
    /** initialize your data structure here. */
    stack<int> DataStack,SortedStack;
    MinStack() {
        while (!DataStack.empty()) {
            DataStack.pop();
        }
        while (!SortedStack.empty()) {
            SortedStack.pop();
        }
    }
    
    void push(int x) {
        DataStack.push(x);
        if (!SortedStack.empty()){
            if(x<=SortedStack.top()){
                SortedStack.push(x);
            }  
        } else {
            SortedStack.push(x);
        }
    }
    
    void pop() {
        if(DataStack.top() == SortedStack.top()){
            SortedStack.pop();
        }
        DataStack.pop();
        return;
    }
    
    int top() {
        int ele = DataStack.top();
        return ele;
    }
    
    int min() {
        if(DataStack.top() <= SortedStack.top()){
            return DataStack.top();
        } 
        else{
            return SortedStack.top();
        }
    }
};

int main(){
    MinStack stack;
    //obj.SQueue();
    stack.push(-2);
    stack.push(0);
    stack.push(-3);
    
    int ele;
    printf("Got %d, expected -3\n", ele = stack.min());
    assert(ele == -3);
    stack.pop();
    printf("Got %d, expected -2\n", ele = stack.min());
    assert(ele == -2);
    printf("Got %d, expected 0\n", ele = stack.top());
    assert(ele == 0);
    stack.pop();
    printf("Got %d, expected -2\n", ele = stack.top());
    assert(ele == -2);    
    printf("Test Completed!\n");
    return 0;
}
